src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs

```diff
@@ -260,6 +259,151 @@ public override void VisitEndDirectiveBlock(DirectiveChunkGenerator chunkGenerat
+            {
+                var renderedBoundAttributeNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                foreach (var attribute in attributes)
+                {
+                    var attributeValueNode = attribute.Value;
+                    var associatedDescriptors = descriptors.Where(descriptor =>
+                        descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Name)));
+
+                    if (associatedDescriptors.Any() && renderedBoundAttributeNames.Add(attribute.Name))
+                    {
+                        if (attributeValueNode == null)
+                        {
+                            // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
+                            // logged an error if it was a bound attribute; so we can skip.
+                            continue;
+                        }
+
+                        foreach (var associatedDescriptor in associatedDescriptors)
+                        {
+                            var associatedAttributeDescriptor = associatedDescriptor.Attributes.First(
+                                attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Name));
+                            var setTagHelperProperty = new SetTagHelperPropertyIRNode()
+                            {
+                                PropertyName = associatedAttributeDescriptor.PropertyName,
+                                AttributeName = attribute.Name,
+                                TagHelperTypeName = associatedDescriptor.TypeName,
+                                SourceRange = BuildSourceRangeFromNode(attributeValueNode),
+                                AssociatedDescriptor = associatedAttributeDescriptor,
+                                ValueStyle = attribute.ValueStyle
+                            };
+
+                            Builder.Push(setTagHelperProperty);
+                            attributeValueNode.Accept(this);
+                            Builder.Pop();
+                        }
+                    }
+                    else
+                    {
+                        var addHtmlAttribute = new AddTagHelperHtmlAttributeIRNode()
+                        {
+                            Name = attribute.Name,
+                            ValueStyle = attribute.ValueStyle
+                        };
+
+                        Builder.Push(addHtmlAttribute);
+                        if (attributeValueNode != null)
+                        {
+                            attributeValueNode.Accept(this);
+                        }
+                        Builder.Pop();
+                    }
+                }
+            }
+
+            private void AddExecuteTagHelpers()
+            {
+                Builder.Add(new ExecuteTagHelpersIRNode());
```

Comment:
I think in the prototype I had this as a marker. We could probably represent this in a better way. Fine for now though; just a design note.